### DISCUSSION TITLE
Next taking +10 seconds to render the page

### DISCUSSION THREAD

**Discussion by joaofmartinho** [vercel/next.js#64225] [Help]
### Summary

I have a simple Next project with one page where I fetch one API and display the data on the screen.
In my use case, the API sends a huge list of items with no pagination (about 3k items grouped). I then iterate these items, and for each group, I create a table and display them all.

My problem is that the page takes 10 seconds to load. (in prod build it takes 4 to 5 seconds)

I can understand that it is a bad case to have so many elements and loops in the component, but at the same time how can another frameworks show the page in  1-2 seconds, sometimes even faster?
Would love some help to know if this is expected or if there is something I am doing wrong.

Here is an example with a simplified use case I extracted from my code:
https://github.com/joaofmartinho/slow-nextjs-case

**Comment by icyJoseph** [comment]
I am not claiming that this will help performance, if after all, the main bottleneck might be awaiting the large list.

I think the `MatchesListTables`, has to be a server component, and then map over the matches, already rendering JSX:

```tsx
export const MatchesListTables = ({ matches }: { matches: TLeagueData[] }) => {

	return matches.map((match) => {
		return <Fragment key={match.league_id}>
			<LeagueInfo league={match} /> // You need to define this component
			<GamesInfo matches={match.games} /> // You need to define this component
		</Fragment>
	})
}
```

That'll at least, make things more readable. Again, it should have next to none impact on the result, since the bottleneck might just be the awaiting that happens at the very top.

**Reply by joaofmartinho** [reply]
Yes, the performance is still the same since I have it like that on my working repo, I did it here in just one component to simplify.

Seems like the performance problem is because of the big prop object or the number of elements returned but I don't understand why another framework with SSR can return the document in one second.

**Reply by icyJoseph** [reply]
I mean your are spreading data, and that can already cause a tremendous amount of re-work. 

- https://jpcamara.com/2023/03/07/making-tanstack-table.html

That being said, You can also map the object: 

- https://nextjs.org/blog/security-nextjs-server-components-actions#component-level-data-access

And extract only what's necessary.

**Reply by joaofmartinho** [reply]
Before all thank you so much for helping me understand this.
I removed the unnecessary spread, but the loading time did not decrease. Also, I don't think I can pass less data to the tables component since most of the data comes from the `games` property.

**Comment by joaofmartinho** [comment]
Second use case, I have this other page:

![Screenshot 2024-04-09 at 19 06 33](https://github.com/vercel/next.js/assets/47006444/46680527-8fee-4a99-afee-c3462e9629dc)

There is a big performance hit when I don't destruct my response and pass the full data into components. Page was taking 4 seconds to load, with the destructed data now takes less than 1 second.

**Reply by joaofmartinho** [reply]
All those 3 components are small client components that just display data from the object on the screen from the selected properties that I destructed.

**Reply by joaofmartinho** [reply]
Prod builds of
Next:

https://github.com/vercel/next.js/assets/47006444/df90f845-ec43-4d51-8a90-7b1be6de03fa



Remix:

https://github.com/vercel/next.js/assets/47006444/9187245d-a6ef-4ef7-8598-102c4fc1e6b7

**Comment by exneval** [comment]
Curious about the performance when using page router compare to app router in v14, would you like to try it? while still waiting for devs inspection

**Reply by joaofmartinho** [reply]
This is a copy of an existing PHP website that can return a big list of results way faster fast, and I'm mostly curious about Next App router /RSC performance vs other frameworks (as seen above, Remix) vs the original PHP website. Not very interested currently in trying it with pages / getServerSideProps.

**Comment by eero123456** [comment]
Have you looked at the response? It's huge, like 10MB. And the first 2 megabytes are already basically the RSC serializiation of the table ( which obviously takes quite a lot time to serialize on server and deserialize and render on client ) On my local machine it takes 4 seconds for any actual content to start flowing in and that's after the RSC serialization.

![image](https://github.com/vercel/next.js/assets/156773195/9fbf2fe8-ef75-40ee-811b-831f920e32bd)


If you are rendering large html pages with React, better just use reactdomserver's render methods directly. That's likely what Remix does quite directly and why it's faster.

**Reply by joaofmartinho** [reply]
Thank you for taking with time this!

Yes, this is an extreme example where I said in the description the endpoint returned a lot of data and a lot of tables were rendered and I'm aware it's a bad thing by defaut. But since I was trying to mimic an existing PHP app that returns the list in one second and another framework too, I was curious about why Next takes 4 seconds as opposed to the other two.

At the same time, I am also trying to figure out why destructing those props on the second case makes the page 10x faster, while in Remix it's the same if I pass the full object or not. On the second case the page is small, but because the prop object is big it causes the slowness.

**Comment by joaofmartinho** [comment]
Bumping this since I really would like to have other comments on If I could optimize the render time in any way w/o virtualize the tables

**Reply by icyJoseph** [reply]
Nope, though next@14.2.1, does perform better in the https://github.com/eknkc/ssr-benchmark from x14 to x12 ~ not great yet hehe

**Reply by joaofmartinho** [reply]
I will try to update to check the differences!
Edit: Didn't notice any difference, unfortunately.

**Reply by icyJoseph** [reply]
Yeah, I don't think that release was aiming to fix this performance issue, rather, they had done some improvements in other areas, that made it slightly better, but that's about it for now.

**Reply by joaofmartinho** [reply]
I think I will close the discussion then. Maybe @leerob or some one that knows better the framework inside can give an additional tip.

**Reply by icyJoseph** [reply]
Yes, I'm also monitoring around to see some comments from core maintainers. There's a reddit thread also, I'll check it later today.
